flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java

```diff
@@ -501,6 +501,14 @@ private <P extends ResponseBody> CompletableFuture<P> submitRequest(
                     }
                 });
 
+        // [FLINK-32583] If connectFuture failed instantly but channelFuture is unresolved, it may
+        // mean the executor service Netty is using has shut down, in which case the above listener
+        // to complete channelFuture will never run
+        if (connectFuture.isDone() && !connectFuture.isSuccess() && !channelFuture.isDone()) {
+            channelFuture.completeExceptionally(
+                    new IOException("Netty connection failed instantly. Shutting down?"));
+        }
```

Comment:
Realistically the `channelFuture.isDone()` check will always pass, as it's unlikely the listener would ever have had time to run—I can remove it if a reviewer wishes.

Also, I'm open to suggestions for a different exception to use than IOException.

Reply:
`IOException` seems to be reasonable. At least, I can't come up with a better alternative :thinking: